### PR TITLE
Make it possible to dump zig files in all modes

### LINKED ISSUE #7844: Dump the zig files in dev/test modes
Currently, `-Dquarkus.debug.generated-classes-dir` only dumps the bytecode. It would be useful to dump the zig files to ease the analysis of the stack trace: 
```
2020-03-13 11:26:24,534 SEVERE [io.ver.cor.imp.DeploymentManager] (vert.x-eventloop-thread-7) Undeploy failed: java.lang.NullPointerException
	at io.quarkus.vertx.verticles.MyBeanVerticle_ClientProxy.arc$delegate(MyBeanVerticle_ClientProxy.zig:442)
	at io.quarkus.vertx.verticles.MyBeanVerticle_ClientProxy.stop(MyBeanVerticle_ClientProxy.zig:36)
	at io.vertx.core.impl.DeploymentManager$DeploymentImpl.lambda$doUndeploy$6(DeploymentManager.java:731)
```

### PR DESCRIPTION
**Pull request by mkouba**
- resolves #7844

### CODE REVIEW

**Comment by mkouba** [comment]
The dump is activated by a system property: `-Dquarkus.debug.generated-sources-dir=/path/to/dir` (aligned with `quarkus.debug.generated-classes-dir`).

File: core/deployment/src/main/java/io/quarkus/deployment/GeneratedClassGizmoAdaptor.java [hunk 1]
```diff
@@ -1,22 +1,48 @@
...
+    public void write(String className, byte[] bytes) {
+        String source = null;
+        if (sources != null) {
+            StringWriter sw = sources.get(className);
+            if (sw != null) {
+                source = sw.toString();
+            }
+        }
+        classOutput.produce(
+                new GeneratedClassBuildItem(applicationClass, className, bytes, source));
```
Comment (gastaldi) [inline core/deployment/src/main/java/io/quarkus/deployment/GeneratedClassGizmoAdaptor.java]: Should this be invoked if `source == null` ?

Comment (mkouba) [reply core/deployment/src/main/java/io/quarkus/deployment/GeneratedClassGizmoAdaptor.java]: Yes, it's ok to pass `null`. It just means "no source available".

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ComponentsProviderGenerator.java [hunk 2]
```diff
@@ -123,7 +124,7 @@ public ComponentsProviderGenerator(AnnotationLiteralProcessor annotationLiterals
         for (Resource resource : classOutput.getResources()) {
             resources.add(resource);
             resources.add(ResourceImpl.serviceProvider(ComponentsProvider.class.getName(),
-                    (resource.getName().replace('/', '.')).getBytes(Charset.forName("UTF-8"))));
+                    (resource.getName().replace('/', '.')).getBytes(Charset.forName("UTF-8")), null));
```
Comment (gastaldi) [inline independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ComponentsProviderGenerator.java]:
```suggestion
                    (resource.getName().replace('/', '.')).getBytes(StandardCharsets.UTF_8), null));
```

**Review by gastaldi — APPROVED** [review]
LGTM

**Review by geoand — APPROVED** [review]
Went through the code and it looks good to me. Also tried the PR and it works great.

**Review by gastaldi — COMMENTED** [review]
It would be nice to have this feature documented somewhere too

**Comment by mkouba** [comment]
> It would be nice to have this feature documented somewhere too.

+1 `quarkus.debug.generated-classes-dir` should be documented as well.

Feel free to add something to this PR.

**Review by manovotn — APPROVED** [review]
Tried locally and works like a charm, I didn't even know this was a thing until today. Good stuff! :-)

**Review by stuartwdouglas — APPROVED** [review]

**Pull request merged**